In [3]:
!pip install dask
!pip install pandas
!pip install psutil

In [23]:
!pip install kaggle pandas zipfile36 --quiet

In [24]:
from google.colab import files
import os

print("📤 سيتم فتح نافذة لرفع ملف kaggle.json...")
uploaded = files.upload()

📤 سيتم فتح نافذة لرفع ملف kaggle.json...


Saving kaggle.json to kaggle.json


In [25]:
import os
os.makedirs('/root/.kaggle', exist_ok=True)
for filename in uploaded.keys():
    os.rename(filename, f'/root/.kaggle/{filename}')
    print(f"✅ تم نسخ {filename} إلى /root/.kaggle/")
!chmod 600 /root/.kaggle/kaggle.json
print("✅ تم تعيين تصاريح الملف بشكل آمن")

✅ تم نسخ kaggle.json إلى /root/.kaggle/
✅ تم تعيين تصاريح الملف بشكل آمن


In [42]:
import os, zipfile


!kaggle datasets download -d mkechinov/ecommerce-behavior-data-from-multi-category-store


with zipfile.ZipFile("ecommerce-behavior-data-from-multi-category-store.zip", "r") as z:
    z.extractall(".")

Dataset URL: https://www.kaggle.com/datasets/mkechinov/ecommerce-behavior-data-from-multi-category-store
License(s): copyright-authors
 99% 4.23G/4.29G [04:13<00:04, 16.2MB/s]
100% 4.29G/4.29G [04:13<00:00, 18.2MB/s]


In [45]:
import pandas as pd
import time
import psutil
import os

csv_path = "/content/2019-Nov.csv"

start_time = time.time()
chunk_size = 5000
chunks = []
process = psutil.Process(os.getpid())

print("🚀 بدء القراءة على دفعات...")

for i, chunk in enumerate(pd.read_csv(csv_path, chunksize=chunk_size)):
    chunks.append(len(chunk))
    print(f"✅ دفعة {i+1}: {len(chunk)} صفوف | الذاكرة الحالية: {process.memory_info().rss / 1024 ** 2:.2f} MB")

print("\n✅ انتهت عملية القراءة على دفعات")
print("📊 مجموع الصفوف المقروءة:", sum(chunks))
print("⏱ الوقت المستغرق:", round(time.time() - start_time, 2), "ثانية")


Streaming output truncated to the last 5000 lines.
✅ دفعة 8506: 5000 صفوف | الذاكرة الحالية: 136.55 MB
✅ دفعة 8507: 5000 صفوف | الذاكرة الحالية: 136.55 MB
✅ دفعة 8508: 5000 صفوف | الذاكرة الحالية: 136.55 MB
✅ دفعة 8509: 5000 صفوف | الذاكرة الحالية: 136.55 MB
✅ دفعة 8510: 5000 صفوف | الذاكرة الحالية: 136.55 MB
✅ دفعة 8511: 5000 صفوف | الذاكرة الحالية: 136.55 MB
✅ دفعة 8512: 5000 صفوف | الذاكرة الحالية: 136.55 MB
✅ دفعة 8513: 5000 صفوف | الذاكرة الحالية: 136.55 MB
✅ دفعة 8514: 5000 صفوف | الذاكرة الحالية: 136.55 MB
✅ دفعة 8515: 5000 صفوف | الذاكرة الحالية: 136.55 MB
✅ دفعة 8516: 5000 صفوف | الذاكرة الحالية: 136.55 MB
✅ دفعة 8517: 5000 صفوف | الذاكرة الحالية: 136.55 MB
✅ دفعة 8518: 5000 صفوف | الذاكرة الحالية: 136.55 MB
✅ دفعة 8519: 5000 صفوف | الذاكرة الحالية: 136.55 MB
✅ دفعة 8520: 5000 صفوف | الذاكرة الحالية: 136.55 MB
✅ دفعة 8521: 5000 صفوف | الذاكرة الحالية: 136.55 MB
✅ دفعة 8522: 5000 صفوف | الذاكرة الحالية: 136.55 MB
✅ دفعة 8523: 5000 صفوف | الذاكرة الحالية: 136.55 MB
✅ دفعة 8524: 

In [47]:
import dask.dataframe as dd
import time
import psutil
import os

csv_path = "/content/2019-Nov.csv"

start_time = time.time()

print(" تحميل الملف بواسطة Dask...")
df = dd.read_csv(csv_path)

print(" تم إنشاء Dask DataFrame")
print(" بعض المعلومات:")
print(df.head())


print("\n📊 حساب الإحصاءات الأساسية (describe)...")
summary = df.describe().compute()

print(summary)
print("\n⏱ الوقت المستغرق:", round(time.time() - start_time, 2), "ثانية")
print("💾 الذاكرة الحالية:", psutil.Process(os.getpid()).memory_info().rss / 1024 ** 2, "MB")


 تحميل الملف بواسطة Dask...
 تم إنشاء Dask DataFrame
 بعض المعلومات:
                event_time event_type  product_id          category_id  \
0  2019-11-01 00:00:00 UTC       view     1003461  2053013555631882655   
1  2019-11-01 00:00:00 UTC       view     5000088  2053013566100866035   
2  2019-11-01 00:00:01 UTC       view    17302664  2053013553853497655   
3  2019-11-01 00:00:01 UTC       view     3601530  2053013563810775923   
4  2019-11-01 00:00:01 UTC       view     1004775  2053013555631882655   

               category_code   brand   price    user_id  \
0     electronics.smartphone  xiaomi  489.07  520088904   
1  appliances.sewing_machine  janome  293.65  530496790   
2                       <NA>   creed   28.31  561587266   
3  appliances.kitchen.washer      lg  712.87  518085591   
4     electronics.smartphone  xiaomi  183.27  558856683   

                           user_session  
0  4d3b30da-a5e4-49df-b1a8-ba5943f1dd33  
1  8e5f4f83-366c-4f70-860e-ca7417414283  
2  75

In [48]:
import gzip
import shutil
import time

csv_path = "/content/2019-Nov.csv"
compressed_path = "/content/2019-Nov.csv.gz"

start_time = time.time()
print("🚀 بدء ضغط الملف...")

with open(csv_path, 'rb') as f_in:
    with gzip.open(compressed_path, 'wb') as f_out:
        shutil.copyfileobj(f_in, f_out)

print("✅ تم ضغط الملف بنجاح.")
print("⏱ الوقت المستغرق للضغط:", round(time.time() - start_time, 2), "ثانية")


🚀 بدء ضغط الملف...
✅ تم ضغط الملف بنجاح.
⏱ الوقت المستغرق للضغط: 847.19 ثانية


In [50]:
import pandas as pd
import psutil
import time

start_time = time.time()
print("🚀 قراءة الملف المضغوط...")

df_compressed = pd.read_csv(compressed_path, compression='gzip', nrows=100000)

print("✅ تم قراءة العينة من الملف المضغوط.")
print("📊 حجم العينة:", df_compressed.shape)
print("⏱ الوقت:", round(time.time() - start_time, 2), "ثانية")
print("💾 الذاكرة:", psutil.Process().memory_info().rss / 1024 ** 2, "MB")


🚀 قراءة الملف المضغوط...
✅ تم قراءة العينة من الملف المضغوط.
📊 حجم العينة: (100000, 9)
⏱ الوقت: 0.4 ثانية
💾 الذاكرة: 540.1640625 MB


# **مقارنة طرق معالجة ملف كبير - النتائج المحدثة**

## **جدول المقارنة المحدث**

| الطريقة | الوقت | الذاكرة | الصفوف |
|---------|--------|----------|---------|
| **Pandas + Chunking** | 239.19 ثانية | 136.19 MB | 67,501,979 |
| **Dask** | 203.92 ثانية | 484.90 MB | 67,501,979 |
| **Compression** | 847.59 ثانية | 540.16 MB | 100,000 |

## **الترتيب حسب الأداء المحدث**

### **الأسرع: Dask**
- **الوقت:** 203.92 ثانية
- **الذاكرة:** 484.90 MB
- **كامل البيانات:** 67.5 مليون صف

### ** الثاني: Pandas + Chunking**
- **الوقت:** 239.19 ثانية
- **الذاكرة:** 136.19 MB (الأفضل في استهلاك الذاكرة)
- **كامل البيانات:** 67.5 مليون صف

### ** الثالث: Compression**
- **الوقت الإجمالي:** 847.59 ثانية
- **الذاكرة:** 540.16 MB
- **عينة محدودة:** 100,000 صف فقط

## **الاستنتاجات المحدثة**

1. **Dask أصبح الأسرع لمعالجة الملفات الكبيرة جداً** (67.5 مليون صف)
2. **Pandas + Chunking الأفضل في كفاءة استخدام الذاكرة** (136MB فقط!)
3. **Dask يتفوق في السرعة مع زيادة حجم البيانات** (أسرع بـ 35 ثانية)
4. **Compression غير عملي للملفات الكبيرة** - الوقت مضاعف 4 مرات

## **التوصيات الجديدة**

- **استخدم Dask عندما:** تكون السرعة هي الأولوية والذاكرة متوفرة
- **استخدم Pandas + Chunking عندما:** تكون كفاءة الذاكرة مهمة
- **تجنب Compression ل:** الملفات الكبيرة التي تحتاج معالجة كاملة

## **ملاحظة تقنية**
- **زيادة حجم البيانات من 9.3M إلى 67.5M صف أظهر تفوق Dask في الأداء**
- **Pandas + Chunking حافظ على كفاءة ذاكرة مذهلة رغم حجم البيانات الهائل**
- **الفرق في الوقت أصبح ملحوظاً مع البيانات الأكبر**